packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts

```diff
@@ -0,0 +1,136 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { Route } from '@tupaia/server-boilerplate';
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityType } from '../../../models';
+import { formatEntitiesForResponse } from '../format';
+import { RelationsRequest } from './types';
+
+type Pair = {
+  descendant: string;
+  ancestor: string;
+};
+
+export class EntityRelationsRoute extends Route<RelationsRequest> {
+  async buildAncestorCodesAndPairs(descendants: EntityType[]): Promise<[string[], Pair[]]> {
+    const { hierarchyId } = this.req.ctx;
+    const { type: ancestorType } = this.req.ctx.ancestor;
+    const descendantAncestorMapping = await this.req.models.entity.fetchAncestorDetailsByDescendantCode(
+      descendants.map(descendant => descendant.code),
+      hierarchyId,
+      ancestorType,
+    );
+    const ancestorCodes = Object.values(descendantAncestorMapping).map(ancestor => ancestor.code);
+    const ancestorDescendantPairs = Object.entries(descendantAncestorMapping).map(
+      ([descendant, { code: ancestor }]) => ({
+        descendant,
+        ancestor,
+      }),
+    );
+
+    return [ancestorCodes, ancestorDescendantPairs];
+  }
+
+  async getAncestors(ancestorsWithDescendantsCodes: string[]) {
+    const { entity, hierarchyId, allowedCountries } = this.req.ctx;
+    const { type: ancestorType, filter } = this.req.ctx.ancestor;
+    return entity.type === ancestorType
+      ? [entity]
+      : entity.getDescendants(hierarchyId, {
+          code: ancestorsWithDescendantsCodes,
+          ...filter,
+          country_code: allowedCountries,
+        });
+  }
+
+  async getFormattedEntitiesByCode(ancestors: EntityType[], descendants: EntityType[]) {
+    const { models, ctx } = this.req;
+    const { flat: ancestorFlat } = ctx.ancestor;
+    const { flat: descendantFlat } = ctx.descendant;
+
+    const formattedEntities =
+      ancestorFlat === descendantFlat
+        ? await formatEntitiesForResponse(models, ctx, [...ancestors, ...descendants], ancestorFlat)
+        : [
+            ...(await formatEntitiesForResponse(models, ctx, ancestors, ancestorFlat)),
+            ...(await formatEntitiesForResponse(models, ctx, descendants, descendantFlat)),
+          ];
+    const formattedEntitiesByCode: Record<string, string> = {};
+    [...ancestors, ...descendants].forEach((entity, index) => {
+      formattedEntitiesByCode[entity.code] = formattedEntities[index];
```

Comment:
I guess that `formatEntitiesForResponse` guarantees a 1-1 and ordered mapping between the entity input and output?

Reply:
Correct